### ISSUE TITLE
`as of` queries fail in StarRocks

### ISSUE THREAD

**Issue by timsehn** [dolthub/dolt#4707]
Customer is trying to use StarRocks DB:

https://github.com/StarRocks/starrocks

They are able connect but apparently `as of` queries don't work. This might need a change to their code.

**Comment by timsehn** [comment]
This may be related:
```
tmp> call dolt_commit('-Am', 'msg');
+----------------------------------+
| hash                             |
+----------------------------------+
| 2vihjhjblv12sa58cl5c5iqp0fu1gj1f |
+----------------------------------+
1 row in set (0.00 sec)

tmp> select * from t as of '2022-11-4';
branch not found: 2022-11-4
tmp> select * from t as of '2022-11-5';
branch not found: 2022-11-5
tmp> select * from t as of now();
Empty set (0.00 sec)
tmp> select * from t as of date('2022-11-4');
branch not found: 2022-11-04
tmp>  
```

**Comment by zachmu** [comment]
Not sure it this is related, but if the `AS OF` query is being generated, it could fail because Dolt's `AS OF` syntax is slightly non standard.

The standard is `select * from table for system_time as of ...`. Dolt omits "for system_time", and won't even parse it. This is due to a thorny conflict in the grammar we have never fixed.

**Comment by timsehn** [comment]
Installing StarRocks on my Mac using:

https://docs.starrocks.io/en-us/latest/quick_start/Deploy

**Comment by timsehn** [comment]
From customer:

```
According to your docs, you can use a timestamp, so I am using the first option in the list below

SELECT * FROM myTable AS OF TIMESTAMP("2020-03-01");
SELECT * FROM myTable AS OF 'HEAD~';
SELECT * FROM myTable AS OF 'HEAD~20';
SELECT * FROM myTable AS OF 'feature-branch';
SELECT * FROM myTable AS OF '4gunscn1j6ijtsj121m7bjaj4j8grjmv';
```

**Comment by timsehn** [comment]
After we get it running I think we need to use this:

https://docs.starrocks.io/en-us/latest/data_source/External_table#external-table-for-a-jdbc-compatible-database

to get it connected to dolt.

**Comment by timsehn** [comment]
Update here: We tried to get StarRocks running on Friday to no avail. I'll let @coffeegoddd give an update on how far h got.

**Comment by coffeegoddd** [comment]
I followed the StarRocks Quick Start guide [here](https://docs.starrocks.io/en-us/latest/quick_start/Deploy), trying to get it deployed on a single EC2 host running Centos 7.

I saw the log messages in the FE and BE that are supposed to indicate a successful start, but never got [step 4](https://docs.starrocks.io/en-us/latest/quick_start/Deploy#step-4-set-up-a-starrocks-cluster) working... I couldn't get the mysql client to connect in order to run the SQL demonstrating a working cluster.

**Comment by chrisfw** [comment]
Hi @coffeegoddd  and @timsehn 

Below are the notes I took from my install experience.  Let me know if you need additional assistance.
Thanks,
Chris

```


https://docs.starrocks.io/en-us/latest/quick_start/Deploy

Set up starrocks

#may have to do this 
mkfs -t ext4 /dev/sdb
mkdir /data
mount /dev/sdb /data

Install Java 13 from rpm
rpm -ivh jdk-13.0.2_linux-x64_bin.rpm

Check lsblk and add a starrocks os user with home dir set to large disk volume
useradd -b /data starrocks

Move the Starrocks install gzip to the starrocks home dir and unzip it
mv StarRocks-2.3.0.tar.gz /data/starrocks/
chown starrocks:starrocks ~starrocks/StarRocks-2.3.0.tar.gz
su - starrocks
tar zxvf StarRocks-2.3.0.tar.gz

set JAVA_HOME in conf (this must be done for be.conf and fe.conf and set for the broker in .bashrc)
JAVA_HOME = /usr/java/jdk-13.0.2


install the mysql client
yum install mysql

#Timeout error is likely caused by a port being blocked by the firewall
#ERROR 1064 (HY000): fail to create tablet: timed out. unfinished replicas(3/10): 76058(10.18.107.24) 76066(10.18.107.24) 76054(10.18.107.24)  timeout=5s


#FE Node - 8030, 8060, 9000, 9010, 9020, 9030
firewall-cmd --zone=public --permanent --add-port=8030/tcp
firewall-cmd --zone=public --permanent --add-port=8060/tcp
firewall-cmd --zone=public --permanent --add-port=9000/tcp
firewall-cmd --zone=public --permanent --add-port=9010/tcp
firewall-cmd --zone=public --permanent --add-port=9020/tcp
firewall-cmd --zone=public --permanent --add-port=9030/tcp
firewall-cmd --reload


--dolt
firewall-cmd --zone=public --permanent --add-port=3306/tcp

#mysql port
firewall-cmd --zone=public --permanent --add-port=9030/tcp
#broker port (put on fe sds node because it has more resources)
firewall-cmd --zone=public --permanent --add-port=8000/tcp
firewall-cmd --reload

Prepare the FE node
vi /etc/security/limits.conf
starrocks        hard    nofile          75000
 
mkdir -p fe/meta
Start the node

Hit the UI in the browser (user is root password is empty string)
Need OCI ingress rules for web ui and minio 
http://10.1.6.210:8030/

wget --user root --password ""  10.18.107.5:8030/

mysql -h 127.0.0.1 -P9030 -uroot
# \q to quit the mysql shell


#on BE server  8030, 8040, 8060, 9000,9050, 9060
firewall-cmd --zone=public --permanent --add-port=8030/tcp
firewall-cmd --zone=public --permanent --add-port=8040/tcp
firewall-cmd --zone=public --permanent --add-port=8060/tcp
firewall-cmd --zone=public --permanent --add-port=9000/tcp
firewall-cmd --zone=public --permanent --add-port=9010/tcp
firewall-cmd --zone=public --permanent --add-port=9020/tcp
firewall-cmd --zone=public --permanent --add-port=9050/tcp
#9060 is critical for cluster table creation
firewall-cmd --zone=public --permanent --add-port=9060/tcp  
firewall-cmd --reload

#make the storage directory
mkdir -p be/storage
#In be.conf set:
storage_root_path = ${STARROCKS_HOME}/storage

#Increase the ulimit
ulimit -n 65535

#Change the limit in /etc/security/limits.conf
starrocks        hard    nofile          75000


#on FE node in mysql compatible front end
SHOW PROC '/frontends'\G
ALTER SYSTEM ADD BACKEND "10.1.0.26:9050";
SHOW PROC '/backends'\G
ALTER SYSTEM decommission BACKEND "10.1.0.26:9050";

bin/start_be.sh --daemon
mysql -h 10.1.6.210 -P9030 -uroot

Set up broker
set JAVA_HOME in conf (this must also be done for be.conf and fe.conf)
for the broker, I added it to .bashrc
export JAVA_HOME=/usr/java/jdk-13.0.2

#needs port 8000 open for cluster access
./apache_hdfs_broker/bin/start_broker.sh --daemon

#add to cluster from fe node (sds db server)
ALTER SYSTEM ADD BROKER broker1 "10.1.0.26:8000";

SHOW PROC '/brokers'\G
```

**Comment by timsehn** [comment]
Thanks. We'll try again with this info.

**Comment by coffeegoddd** [comment]
@chrisfw thanks for the detailed notes! 

I was able to get StarRocks running successfully with an external Dolt datasource. I can confirm that the errors from StarRocks are SQL parsing errors, suggesting they don't support the `AS OF` syntax.

```
2022-11-10 00:45:36,618 WARN (starrocks-mysql-nio-pool-16|258) [ConnectProcessor.handleQuery():304] Process one query failed because.
com.starrocks.common.AnalysisException: Syntax error in line 1:
...m dolt_corona_virus as of timestamp('2020-03-09') limi...
                             ^
Encountered: TIMESTAMP
Expected: COMMA

	at com.starrocks.qe.ConnectProcessor.handleQuery(ConnectProcessor.java:266) ~[starrocks-fe.jar:?]
	at com.starrocks.qe.ConnectProcessor.dispatch(ConnectProcessor.java:401) ~[starrocks-fe.jar:?]
	at com.starrocks.qe.ConnectProcessor.processOnce(ConnectProcessor.java:637) ~[starrocks-fe.jar:?]
	at com.starrocks.mysql.nio.ReadListener.lambda$handleEvent$0(ReadListener.java:55) ~[starrocks-fe.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:830) [?:?]
```

```
2022-11-10 00:47:53,823 WARN (starrocks-mysql-nio-pool-17|263) [ConnectProcessor.handleQuery():304] Process one query failed because.
com.starrocks.common.AnalysisException: Syntax error in line 1:
select * from dolt_corona_virus as of 'alt' limit 10
                                      ^
Encountered: STRING LITERAL
Expected: COMMA

	at com.starrocks.qe.ConnectProcessor.handleQuery(ConnectProcessor.java:266) ~[starrocks-fe.jar:?]
	at com.starrocks.qe.ConnectProcessor.dispatch(ConnectProcessor.java:401) ~[starrocks-fe.jar:?]
	at com.starrocks.qe.ConnectProcessor.processOnce(ConnectProcessor.java:637) ~[starrocks-fe.jar:?]
	at com.starrocks.mysql.nio.ReadListener.lambda$handleEvent$0(ReadListener.java:55) ~[starrocks-fe.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:830) [?:?]
```

Looks like you've already opened a ticket there 🤠 https://github.com/StarRocks/starrocks/issues/10004 .

**Comment by chrisfw** [comment]
No problem Dustin.  I am glad you were able to reproduce the issue.  Tim is aware from an earlier call we had that the issue is in their connector code but said he would be willing to have your team investigate and attempt to update the StarRocks connector code to work properly with Dolt.Let me know if you need any more info.  Thanks for your help with this!Regards,ChrisOn Nov 9, 2022, at 8:12 PM, Dustin Brown ***@***.***> wrote:﻿
@chrisfw thanks for the detailed notes!
I was able to get StarRocks running successfully with an external Dolt datasource. I can confirm that the errors from StarRocks are SQL parsing errors, suggesting they don't support the AS OF syntax.
2022-11-10 00:45:36,618 WARN (starrocks-mysql-nio-pool-16|258) [ConnectProcessor.handleQuery():304] Process one query failed because.
com.starrocks.common.AnalysisException: Syntax error in line 1:
...m dolt_corona_virus as of timestamp('2020-03-09') limi...
                             ^
Encountered: TIMESTAMP
Expected: COMMA

	at com.starrocks.qe.ConnectProcessor.handleQuery(ConnectProcessor.java:266) ~[starrocks-fe.jar:?]
	at com.starrocks.qe.ConnectProcessor.dispatch(ConnectProcessor.java:401) ~[starrocks-fe.jar:?]
	at com.starrocks.qe.ConnectProcessor.processOnce(ConnectProcessor.java:637) ~[starrocks-fe.jar:?]
	at com.starrocks.mysql.nio.ReadListener.lambda$handleEvent$0(ReadListener.java:55) ~[starrocks-fe.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:830) [?:?]

2022-11-10 00:47:53,823 WARN (starrocks-mysql-nio-pool-17|263) [ConnectProcessor.handleQuery():304] Process one query failed because.
com.starrocks.common.AnalysisException: Syntax error in line 1:
select * from dolt_corona_virus as of 'alt' limit 10
                                      ^
Encountered: STRING LITERAL
Expected: COMMA

	at com.starrocks.qe.ConnectProcessor.handleQuery(ConnectProcessor.java:266) ~[starrocks-fe.jar:?]
	at com.starrocks.qe.ConnectProcessor.dispatch(ConnectProcessor.java:401) ~[starrocks-fe.jar:?]
	at com.starrocks.qe.ConnectProcessor.processOnce(ConnectProcessor.java:637) ~[starrocks-fe.jar:?]
	at com.starrocks.mysql.nio.ReadListener.lambda$handleEvent$0(ReadListener.java:55) ~[starrocks-fe.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:830) [?:?]

Looks like you've already opened a ticket there 🤠 StarRocks/starrocks#10004 .

**Comment by fulghum** [comment]
Hey @chrisfw, just a quick update that I'm digging into the StarRocks code and I can confirm their parser/grammar doesn't currently have support for temporal queries with `AS OF`. I've got some changes prototyped locally that add support for that, but I still need to do some more end-to-end testing with Dolt and make sure it's really working correctly. I'm planning to make more progress on that today and get a PR opened with StarRocks tomorrow. 

While testing the StarRocks sql grammar, I noticed that [another form of temporal queries that Dolt supports](https://docs.dolthub.com/sql-reference/version-control/querying-history#specifying-a-commit-hash-for-an-entire-connection) seems to work with the current StarRocks parser. This will only work with a Dolt commit ID or commit reference, and I believe you are looking for timestamp support, so this probably doesn't do exactly what you're looking for anyway, but figured I'd pass this along just in case it helps. 

```
SELECT * FROM `mydb/HEAD~`.myTable;
SELECT * FROM `mydb/HEAD~20`.myTable;
SELECT * FROM `mydb/feature-branch`.myTable;
SELECT * FROM `mydb/4gunscn1j6ijtsj121m7bjaj4j8grjmv`.myTable;
```

I'll keep you updated as my testing progresses.

**Comment by chrisfw** [comment]
@fulghum , thanks so much for digging into the Starrocks code and working on implementing the ```AS OF``` functionality with the timestamp.  I really appreciate it.  Thanks also for the heads up on the other form of temporal queries, that is good to know and may prove useful in other scenarios.  You are correct though in that we are looking to use the date rather than a particular commit id or commit reference.

Please let me know if you need anything from me.

Regards,
Chris

**Comment by fulghum** [comment]
My pleasure! I made some more progress today, but I realized there's more code that needs to be wired together for this to work correctly end-to-end. The StarRocks FE nodes can now accept the temporal query syntax, but they just drop it on the floor and don't pass it to the BE nodes yet. 

It looks like there's some work in the FE code to pass the temporal clause around, some updates to the Thrift RPC messages so the FE node can send it to the BE node, and then an update in the BE node code to actually send the temporal clause as part of the query to the external DB. (I'm currently testing with the MySQL external table support, but I believe the JDBC external table support should be very similar.) Setting up the BE C++ build slowed me down a little today, but I'm still making progress and will keep this thread updated.

**Comment by chrisfw** [comment]
Thanks for the update @fulghum , it is really good to hear that you continue to make progress.  I'm sorry that the implementation has required more extensive updates than one would hope.  It will be really awesome to see this in action once complete!  

Regards,
Chris

**Comment by fulghum** [comment]
I needed to prioritize working on an analyzer bug fix yesterday, but I did take a little time to test out my StarRocks changes, and it looks like this is working correctly now. Some sample query results below. I'll get this cleaned up and try to get a PR opened with StarRocks later today. 

```sql
mysql> select * from mysql_external_table1;
+------+------------+
| pk   | value      |
+------+------------+
|    1 | one        |
|    2 | twoooo     |
|  100 | one hundy! |
+------+------------+
3 rows in set (0.00 sec)

mysql> select * from mysql_external_table1 as of 't2c0vr0m1o84h28fnr8m6c2fe5oariii';
+------+--------+
| pk   | value  |
+------+--------+
|    1 | one    |
|    2 | twoooo |
+------+--------+
2 rows in set (0.01 sec)

mysql> select * from mysql_external_table1 as of '0pe6o4rg5fkb8jlippjf4s50aht09u50';
+------+-------+
| pk   | value |
+------+-------+
|    1 | one   |
+------+-------+
1 row in set (0.02 sec)

mysql> select * from mysql_external_table1 as of 'does_not_exist';
ERROR 1064 (HY000): mysql query failed. Err: branch not found: does_not_exist
```

**Comment by fulghum** [comment]
PR opened: https://github.com/StarRocks/starrocks/pull/13707

**Comment by chrisfw** [comment]
@fulghum , thank you, that is awesome news!  One question, in you sample above none of the queries use the as if timestamp with a date.  Is that working as well?

Have a great weekend!

Regards,
Chris

**Comment by timsehn** [comment]
<img width="749" alt="image" src="https://user-images.githubusercontent.com/5376236/202812206-01fea6e9-a552-4cde-94e1-7bcffa4b2028.png">

It's in the tests :-)

**Comment by fulghum** [comment]
My pleasure @chrisfw! Sorry to leave out that important example. Support for `as of timestamp('')` is working correctly as well. 😄 Here's a sample query for you: 

```sql
mysql> select * from mysql_external_table1 as of timestamp('2022-11-16 18:59:42');
+------+-------+
| pk   | value |
+------+-------+
|    1 | one   |
+------+-------+
1 row in set (0.02 sec)
```

**Comment by chrisfw** [comment]
This is really great @fulghum , I really appreciate the effort you put in to get this implemented and a big thanks also to @timsehn for making this happen for us!

**Comment by timsehn** [comment]
Looks like we've got positive reviewer feedback so the StarRocks PR is likely to be accepted!

**Comment by fulghum** [comment]
The StarRocks PR (https://github.com/StarRocks/starrocks/pull/13707) was accepted and merged in. 🎉  Very neat feature to be able to use temporal queries with StarRocks to query historical data in external MySQL tables. 

@chrisfw – thank you for working with us so we could add this integration. This should go out in the next StarRocks release, but feel free to ping me if you want to try building StarRocks from source and I'm happy to help.

**Comment by chrisfw** [comment]
@fulghum and @timsehn , this is really great news! Thanks for all of the effort and the initiative to make this happen!  It was my pleasure to assist with this.
I may take you up on your offer to build StarRocks from source with the new functionality - I will definitely reach out.

Thanks again! 
-Chris

**Comment by fulghum** [comment]
Very happy we could get this integration in place! I'll go ahead and resolve this issue, since the PR has been merged into StarRocks; feel free to reach out here or privately if we can help out more.

**Comment by chrisfw** [comment]
Thanks @fulghum , sounds good!
